src/devices/obs.ts

```diff
@@ -0,0 +1,740 @@
+		)
+		this._doOnTime.on('error', (e) =>
+			this.emit('error', 'OBS.doOnTime', e)
+		)
+		this._doOnTime.on('slowCommand', (msg) =>
+			this.emit('slowCommand', this.deviceName + ': ' + msg)
+		)
+	}
+	init (options: OBSOptions): Promise<boolean> {
+		this._options = options
+		this._obs = new OBSWebSocket()
+		this._obs.on('AuthenticationFailure', () => {
+			this._setConnected(true, false)
+		})
+		this._obs.on('AuthenticationSuccess', () => {
+			this._initialized = true
+			this._setConnected(true, true)
+			this.emit('resetResolver')
+		})
+		this._obs.on('ConnectionClosed', () => {
+			this._setConnected(false)
+			this._triggerRetryConnection()
+		})
+		this._obs.on('error' as any, (e) => this.emit('error', 'OBS', e))
+
+		return this._connect()
+			.then((connected) => {
+				if (!connected) {
+					this._triggerRetryConnection()
+				}
+				return connected
+			})
+	}
+
+	private _connect () {
+		return this._obs
+			.connect({
+				address: `${this._options.host}:${this._options.port}`,
+				password: this._options.password
+			})
+			.then(() => {
+				// connected
+				let time = this.getCurrentTime()
+				let state = this._getDefaultState()
+				this.setState(state, time)
+				this._setConnected(true, this._authenticated)
+				return true
+			})
+			.catch((err) => {
+				// connection error
+				this.emit('error', 'OBS', err)
+				this._setConnected(false)
+				return false
+			})
+	}
+
+	private _connectionChanged () {
+		this.emit('connectionChanged', this.getStatus())
+	}
+
+	private _setConnected (connected: boolean, authenticated: boolean = false) {
+		if (
+			this._connected !== connected ||
+			this._authenticated !== authenticated
+		) {
+			this._connected = connected
+			this._authenticated = authenticated
+			this._connectionChanged()
+
+			if (!this._authenticated) {
+				this._initialized = false
+			}
+		}
+	}
+
+	private _triggerRetryConnection () {
+		if (!this._retryConnectTimeout) {
+			this._retryConnectTimeout = setTimeout(() => {
+				this._retryConnection()
+			}, RETRY_TIMEOUT)
+		}
+	}
+	private _retryConnection () {
+		if (this._retryConnectTimeout) {
+			clearTimeout(this._retryConnectTimeout)
+			this._retryConnectTimeout = undefined
+		}
+
+		if (!this.connected && !this._setDisconnected) {
+			this._connect()
+				.then((connected) => {
+					if (!connected) {
+						this._triggerRetryConnection()
+					}
+				})
+				.catch((err) => {
+					this.emit('error', 'OBS retryConnection', err)
+				})
+		}
+	}
+
+	private _getDefaultState (): OBSState {
+		return {
+			currentScene: undefined,
+			previewScene: undefined,
+			currentTransition: undefined,
+			muted: {},
+			recording: undefined,
+			streaming: undefined,
+			scenes: {},
+			sources: {}
+		}
+	}
+
+	/** Called by the Conductor a bit before a .handleState is called */
+	prepareForHandleState (newStateTime: number) {
+		// clear any queued commands later than this time:
+		this._doOnTime.clearQueueNowAndAfter(newStateTime + 0.1)
```

Comment:
why `newStateTime + 0.1`? Won't that cause commands to be duplicated, as they are about to regenerated for `newStateTime`.

Not always duplicated, this could happen:
* First resolve of some future time 500, queue `currentScene = 1`
* prepareForHandleState(500), clears nothing
* Timeline is changed
* New resolve of time 500, queue `currentScene = 2`

So we will be performing a `currentScene = 1` folloewed 0.1ms later by `currentScene = 2`. This will quite possible cause a single frame flash of 1, depending on obs internals.

Reply:
I don't know 😅 I just copied this from the vMix module... So, what you're saying is that it would be better to do this as `.clearQueueNowAndAfter(newStateTime)`? I'm just a babe in the woods.